add multi-OS support and cleaned up package dependencies

jdbirdwell opened the pull request:
Brett,

I have edited the build-cac-enabled-git script to use a different method to identify the OS and modified the list of required packages so that everything builds on, at least, Debian, Ubuntu, CentOS, and Fedora.  I have not changed the README.md file -- that should be (slightly) updated.

I have tested all the steps EXCEPT steps 3 & 10.  I haven't tested 3 & 10 because I'm still waiting for a CAC (!).  I didn't test on Red Hat Enterprise Linux b/c I don't have a key, but it and CentOS should behave the same.  I also didn't test on Cygwin (again, no system to test).  If you still have access to a Cygwin system it would be good to test that & be sure I didn't mess anything up -- I didn't change (intentionally) the Cygwin-specific code.  The behavior on Ubuntu should also be the same as before.  Perhaps someone else can check the RHEL build (or I might be able to impose on a friend...).

The test for a Mac OS X system is for future use -- the problem there being the complexity of the installer and absence of a uniform set of pre-built packages.  Any systems other than those listed are flagged as "unknown".  Both Mac OS X and unknown cause graceful exits.  Other Linux/UNIX systems can easily be added (with some experimentation to determine which packages to install and, possibly, the installer to use) -- the approach I am using to identify the OS is quite general.  See http://linuxmafia.com/faq/Admin/release-files.html for more information.

Once I get a CAC & a CAC-enabled server I can point to, I will test steps 3 & 10, and execution.  I'm using docker for all the tests. (Example, once you have the docker service installed and running: "docker run -ti -v /absolute-path-to-local-dir:/absolute-path-to-dir-in-container ubuntu:latest /bin/bash", followed by "apt-get update; apt-get install build-essential" in the container, followed by ./build-cac-enabled-git {options}.  Substitute debian:latest, centos:latest, rhel:latest, or fedora:latest as needed, and the appropriate installer -- note that rhel requires a key before it will allow yum to install things.)  I used docker and this approach in part because I wanted to be sure the build works when starting with a minimal development environment.

Note that the script does not check to ensure that the development/build tools have been installed -- the previous version doesn't do this, either.  For Debian/Ubuntu this can be done with "apt-get install build-essential"; for CentOS/Fedora/RHEL the command is "yum groupinstall 'Development Tools'" (or dnf instead of yum if dnf exists).  "apt-get/yum/dnf update" should be executed before this.

One more thing to attempt in the future is to use apt-get/yum/dnf to install OpenSC rather than build by source.  The OpenSC folks replied to my issue and said they have no control over the old versions associated with the various distributions.

Doug

jdbirdwell committed:
add multi-OS support and cleaned up package dependencies

bpstahlman commented:
Doug,
Great. Thanks. Perused the changes, and didn't see any issues. Hopefully,
I'll have time to give a slightly closer look and merge the pull request
this weekend. I agree that on platforms providing some sort of "build
essentials" package, the script should install the package automatically.
Let me know if you have any recommendations on the license…

Thanks,
Brett S.
On Nov 11, 2015 4:33 PM, "Doug Birdwell" notifications@github.com wrote:

> Brett,
> 
> I have edited the build-cac-enabled-git script to use a different method
> to identify the OS and modified the list of required packages so that
> everything builds on, at least, Debian, Ubuntu, CentOS, and Fedora. I have
> not changed the README.md file -- that should be (slightly) updated.
> 
> I have tested all the steps EXCEPT steps 3 & 10. I haven't tested 3 & 10
> because I'm still waiting for a CAC (!). I didn't test on Red Hat
> Enterprise Linux b/c I don't have a key, but it and CentOS should behave
> the same. I also didn't test on Cygwin (again, no system to test). If you
> still have access to a Cygwin system it would be good to test that & be
> sure I didn't mess anything up -- I didn't change (intentionally) the
> Cygwin-specific code. The behavior on Ubuntu should also be the same as
> before. Perhaps someone else can check the RHEL build (or I might be able
> to impose on a friend...).
> 
> The test for a Mac OS X system is for future use -- the problem there
> being the complexity of the installer and absence of a uniform set of
> pre-built packages. Any systems other than those listed are flagged as
> "unknown". Both Mac OS X and unknown cause graceful exits. Other Linux/UNIX
> systems can easily be added (with some experimentation to determine which
> packages to install and, possibly, the installer to use) -- the approach I
> am using to identify the OS is quite general. See
> http://linuxmafia.com/faq/Admin/release-files.html for more information.
> 
> Once I get a CAC & a CAC-enabled server I can point to, I will test steps
> 3 & 10, and execution. I'm using docker for all the tests. (Example, once
> you have the docker service installed and running: "docker run -ti -v
> /absolute-path-to-local-dir:/absolute-path-to-dir-in-container
> ubuntu:latest /bin/bash", followed by "apt-get update; apt-get install
> build-essential" in the container, followed by ./build-cac-enabled-git
> {options}. Substitute debian:latest, centos:latest, rhel:latest, or
> fedora:latest as needed, and the appropriate installer -- note that rhel
> requires a key before it will allow yum to install things.) I used docker
> and this approach in part because I wanted to be sure the build works when
> starting with a minimal development environment.
> 
> Note that the script does not check to ensure that the development/build
> tools have been installed -- the previous version doesn't do this, either.
> For Debian/Ubuntu this can be done with "apt-get install build-essential";
> for CentOS/Fedora/RHEL the command is "yum groupinstall 'Development
> Tools'" (or dnf instead of yum if dnf exists). "apt-get/yum/dnf update"
> should be executed before this.
> 
> One more thing to attempt in the future is to use apt-get/yum/dnf to
> install OpenSC rather than build by source. The OpenSC folks replied to my
> issue and said they have no control over the old versions associated with
> the various distributions.
> 
> ## Doug
> 
> You can view, comment on, or merge this pull request online at:
> 
>   https://github.com/bpstahlman/cac-enabled-git-setup/pull/5
> Commit Summary
> - add multi-OS support and cleaned up package dependencies
> 
> File Changes
> - _M_ build-cac-enabled-git
>   https://github.com/bpstahlman/cac-enabled-git-setup/pull/5/files#diff-0
>   (140)
> 
> Patch Links:
> - https://github.com/bpstahlman/cac-enabled-git-setup/pull/5.patch
> - https://github.com/bpstahlman/cac-enabled-git-setup/pull/5.diff
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/bpstahlman/cac-enabled-git-setup/pull/5.

jdbirdwell commented:
Look into the MIT license. It is quite flexible.
Doug

On Thu, Nov 12, 2015, 8:47 PM Brett Stahlman notifications@github.com
wrote:

> Doug,
> Great. Thanks. Perused the changes, and didn't see any issues. Hopefully,
> I'll have time to give a slightly closer look and merge the pull request
> this weekend. I agree that on platforms providing some sort of "build
> essentials" package, the script should install the package automatically.
> Let me know if you have any recommendations on the license…
> 
> Thanks,
> Brett S.
> On Nov 11, 2015 4:33 PM, "Doug Birdwell" notifications@github.com wrote:
> 
> > Brett,
> > 
> > I have edited the build-cac-enabled-git script to use a different method
> > to identify the OS and modified the list of required packages so that
> > everything builds on, at least, Debian, Ubuntu, CentOS, and Fedora. I
> > have
> > not changed the README.md file -- that should be (slightly) updated.
> > 
> > I have tested all the steps EXCEPT steps 3 & 10. I haven't tested 3 & 10
> > because I'm still waiting for a CAC (!). I didn't test on Red Hat
> > Enterprise Linux b/c I don't have a key, but it and CentOS should behave
> > the same. I also didn't test on Cygwin (again, no system to test). If you
> > still have access to a Cygwin system it would be good to test that & be
> > sure I didn't mess anything up -- I didn't change (intentionally) the
> > Cygwin-specific code. The behavior on Ubuntu should also be the same as
> > before. Perhaps someone else can check the RHEL build (or I might be able
> > to impose on a friend...).
> > 
> > The test for a Mac OS X system is for future use -- the problem there
> > being the complexity of the installer and absence of a uniform set of
> > pre-built packages. Any systems other than those listed are flagged as
> > "unknown". Both Mac OS X and unknown cause graceful exits. Other
> > Linux/UNIX
> > systems can easily be added (with some experimentation to determine which
> > packages to install and, possibly, the installer to use) -- the approach
> > I
> > am using to identify the OS is quite general. See
> > http://linuxmafia.com/faq/Admin/release-files.html for more information.
> > 
> > Once I get a CAC & a CAC-enabled server I can point to, I will test steps
> > 3 & 10, and execution. I'm using docker for all the tests. (Example, once
> > you have the docker service installed and running: "docker run -ti -v
> > /absolute-path-to-local-dir:/absolute-path-to-dir-in-container
> > ubuntu:latest /bin/bash", followed by "apt-get update; apt-get install
> > build-essential" in the container, followed by ./build-cac-enabled-git
> > {options}. Substitute debian:latest, centos:latest, rhel:latest, or
> > fedora:latest as needed, and the appropriate installer -- note that rhel
> > requires a key before it will allow yum to install things.) I used docker
> > and this approach in part because I wanted to be sure the build works
> > when
> > starting with a minimal development environment.
> > 
> > Note that the script does not check to ensure that the development/build
> > tools have been installed -- the previous version doesn't do this,
> > either.
> > For Debian/Ubuntu this can be done with "apt-get install
> > build-essential";
> > for CentOS/Fedora/RHEL the command is "yum groupinstall 'Development
> > Tools'" (or dnf instead of yum if dnf exists). "apt-get/yum/dnf update"
> > should be executed before this.
> > 
> > One more thing to attempt in the future is to use apt-get/yum/dnf to
> > install OpenSC rather than build by source. The OpenSC folks replied to
> > my
> > issue and said they have no control over the old versions associated with
> > the various distributions.
> > 
> > ## Doug
> > 
> > You can view, comment on, or merge this pull request online at:
> > 
> > https://github.com/bpstahlman/cac-enabled-git-setup/pull/5
> > Commit Summary
> > - add multi-OS support and cleaned up package dependencies
> > 
> > File Changes
> > - _M_ build-cac-enabled-git
> >   <https://github.com/bpstahlman/cac-enabled-git-setup/pull/5/files#diff-0
> > 
> > (140)
> > 
> > Patch Links:
> > - https://github.com/bpstahlman/cac-enabled-git-setup/pull/5.patch
> > - https://github.com/bpstahlman/cac-enabled-git-setup/pull/5.diff
> > 
> > —
> > Reply to this email directly or view it on GitHub
> > https://github.com/bpstahlman/cac-enabled-git-setup/pull/5.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/bpstahlman/cac-enabled-git-setup/pull/5#issuecomment-156291857
> .

bpstahlman commented:
Doug,
Looks good. One minor stylistic point…  Was there a reason for using grep
for the various tests against system/distro strings? Why not a simple case
statement, or even glob matching within [[ … ]]?  I wondered whether
perhaps the need to run "type sw_vers" led you to convert the case/esac to
an if/else, but even there, glob matching would have sufficed. It's not a
big deal, but there's something about spawning processes for things that
bash (or even sh) can handle natively that perturbs my aesthetic
sensibilities…  ;-)

Thanks,
Brett S.
On Nov 11, 2015 4:33 PM, "Doug Birdwell" notifications@github.com wrote:

> Brett,
> 
> I have edited the build-cac-enabled-git script to use a different method
> to identify the OS and modified the list of required packages so that
> everything builds on, at least, Debian, Ubuntu, CentOS, and Fedora. I have
> not changed the README.md file -- that should be (slightly) updated.
> 
> I have tested all the steps EXCEPT steps 3 & 10. I haven't tested 3 & 10
> because I'm still waiting for a CAC (!). I didn't test on Red Hat
> Enterprise Linux b/c I don't have a key, but it and CentOS should behave
> the same. I also didn't test on Cygwin (again, no system to test). If you
> still have access to a Cygwin system it would be good to test that & be
> sure I didn't mess anything up -- I didn't change (intentionally) the
> Cygwin-specific code. The behavior on Ubuntu should also be the same as
> before. Perhaps someone else can check the RHEL build (or I might be able
> to impose on a friend...).
> 
> The test for a Mac OS X system is for future use -- the problem there
> being the complexity of the installer and absence of a uniform set of
> pre-built packages. Any systems other than those listed are flagged as
> "unknown". Both Mac OS X and unknown cause graceful exits. Other Linux/UNIX
> systems can easily be added (with some experimentation to determine which
> packages to install and, possibly, the installer to use) -- the approach I
> am using to identify the OS is quite general. See
> http://linuxmafia.com/faq/Admin/release-files.html for more information.
> 
> Once I get a CAC & a CAC-enabled server I can point to, I will test steps
> 3 & 10, and execution. I'm using docker for all the tests. (Example, once
> you have the docker service installed and running: "docker run -ti -v
> /absolute-path-to-local-dir:/absolute-path-to-dir-in-container
> ubuntu:latest /bin/bash", followed by "apt-get update; apt-get install
> build-essential" in the container, followed by ./build-cac-enabled-git
> {options}. Substitute debian:latest, centos:latest, rhel:latest, or
> fedora:latest as needed, and the appropriate installer -- note that rhel
> requires a key before it will allow yum to install things.) I used docker
> and this approach in part because I wanted to be sure the build works when
> starting with a minimal development environment.
> 
> Note that the script does not check to ensure that the development/build
> tools have been installed -- the previous version doesn't do this, either.
> For Debian/Ubuntu this can be done with "apt-get install build-essential";
> for CentOS/Fedora/RHEL the command is "yum groupinstall 'Development
> Tools'" (or dnf instead of yum if dnf exists). "apt-get/yum/dnf update"
> should be executed before this.
> 
> One more thing to attempt in the future is to use apt-get/yum/dnf to
> install OpenSC rather than build by source. The OpenSC folks replied to my
> issue and said they have no control over the old versions associated with
> the various distributions.
> 
> ## Doug
> 
> You can view, comment on, or merge this pull request online at:
> 
>   https://github.com/bpstahlman/cac-enabled-git-setup/pull/5
> Commit Summary
> - add multi-OS support and cleaned up package dependencies
> 
> File Changes
> - _M_ build-cac-enabled-git
>   https://github.com/bpstahlman/cac-enabled-git-setup/pull/5/files#diff-0
>   (140)
> 
> Patch Links:
> - https://github.com/bpstahlman/cac-enabled-git-setup/pull/5.patch
> - https://github.com/bpstahlman/cac-enabled-git-setup/pull/5.diff
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/bpstahlman/cac-enabled-git-setup/pull/5.

jdbirdwell commented:
Not really - it was just easier for me b/c I don't do a lot of shell 
coding.  The time required for grep is negligible relative to the time 
for the script, so I felt it wasn't a big deal.  Either way; it won't 
offend me in the least if you want to change it.

Thanks,
Doug

On 11/18/15 9:12 PM, Brett Stahlman wrote:

> Doug,
> Looks good. One minor stylistic point… Was there a reason for using grep
> for the various tests against system/distro strings? Why not a simple case
> statement, or even glob matching within [[ … ]]? I wondered whether
> perhaps the need to run "type sw_vers" led you to convert the case/esac to
> an if/else, but even there, glob matching would have sufficed. It's not a
> big deal, but there's something about spawning processes for things that
> bash (or even sh) can handle natively that perturbs my aesthetic
> sensibilities… ;-)
> 
> Thanks,
> Brett S.
> On Nov 11, 2015 4:33 PM, "Doug Birdwell" notifications@github.com wrote:
> 
> > Brett,
> > 
> > I have edited the build-cac-enabled-git script to use a different method
> > to identify the OS and modified the list of required packages so that
> > everything builds on, at least, Debian, Ubuntu, CentOS, and Fedora. 
> > I have
> > not changed the README.md file -- that should be (slightly) updated.
> > 
> > I have tested all the steps EXCEPT steps 3 & 10. I haven't tested 3 & 10
> > because I'm still waiting for a CAC (!). I didn't test on Red Hat
> > Enterprise Linux b/c I don't have a key, but it and CentOS should behave
> > the same. I also didn't test on Cygwin (again, no system to test). 
> > If you
> > still have access to a Cygwin system it would be good to test that & be
> > sure I didn't mess anything up -- I didn't change (intentionally) the
> > Cygwin-specific code. The behavior on Ubuntu should also be the same as
> > before. Perhaps someone else can check the RHEL build (or I might be 
> > able
> > to impose on a friend...).
> > 
> > The test for a Mac OS X system is for future use -- the problem there
> > being the complexity of the installer and absence of a uniform set of
> > pre-built packages. Any systems other than those listed are flagged as
> > "unknown". Both Mac OS X and unknown cause graceful exits. Other 
> > Linux/UNIX
> > systems can easily be added (with some experimentation to determine 
> > which
> > packages to install and, possibly, the installer to use) -- the 
> > approach I
> > am using to identify the OS is quite general. See
> > http://linuxmafia.com/faq/Admin/release-files.html for more information.
> > 
> > Once I get a CAC & a CAC-enabled server I can point to, I will test 
> > steps
> > 3 & 10, and execution. I'm using docker for all the tests. (Example, 
> > once
> > you have the docker service installed and running: "docker run -ti -v
> > /absolute-path-to-local-dir:/absolute-path-to-dir-in-container
> > ubuntu:latest /bin/bash", followed by "apt-get update; apt-get install
> > build-essential" in the container, followed by ./build-cac-enabled-git
> > {options}. Substitute debian:latest, centos:latest, rhel:latest, or
> > fedora:latest as needed, and the appropriate installer -- note that rhel
> > requires a key before it will allow yum to install things.) I used 
> > docker
> > and this approach in part because I wanted to be sure the build 
> > works when
> > starting with a minimal development environment.
> > 
> > Note that the script does not check to ensure that the development/build
> > tools have been installed -- the previous version doesn't do this, 
> > either.
> > For Debian/Ubuntu this can be done with "apt-get install 
> > build-essential";
> > for CentOS/Fedora/RHEL the command is "yum groupinstall 'Development
> > Tools'" (or dnf instead of yum if dnf exists). "apt-get/yum/dnf update"
> > should be executed before this.
> > 
> > One more thing to attempt in the future is to use apt-get/yum/dnf to
> > install OpenSC rather than build by source. The OpenSC folks replied 
> > to my
> > issue and said they have no control over the old versions associated 
> > with
> > the various distributions.
> > 
> > ## Doug
> > 
> > You can view, comment on, or merge this pull request online at:
> > 
> > https://github.com/bpstahlman/cac-enabled-git-setup/pull/5
> > Commit Summary
> > - add multi-OS support and cleaned up package dependencies
> > 
> > File Changes
> > - _M_ build-cac-enabled-git
> > 
> > https://github.com/bpstahlman/cac-enabled-git-setup/pull/5/files#diff-0
> > (140)
> > 
> > Patch Links:
> > - https://github.com/bpstahlman/cac-enabled-git-setup/pull/5.patch
> > - https://github.com/bpstahlman/cac-enabled-git-setup/pull/5.diff
> > 
> > —
> > Reply to this email directly or view it on GitHub
> > https://github.com/bpstahlman/cac-enabled-git-setup/pull/5.
> 
> —
> Reply to this email directly or view it on GitHub 
> https://github.com/bpstahlman/cac-enabled-git-setup/pull/5#issuecomment-157925720.

bpstahlman commented:
Understood. Thanks.
Brett S.
On Nov 18, 2015 9:15 PM, "Doug Birdwell" notifications@github.com wrote:

> Not really - it was just easier for me b/c I don't do a lot of shell
> coding. The time required for grep is negligible relative to the time
> for the script, so I felt it wasn't a big deal. Either way; it won't
> offend me in the least if you want to change it.
> 
> Thanks,
> Doug
> 
> On 11/18/15 9:12 PM, Brett Stahlman wrote:
> 
> > Doug,
> > Looks good. One minor stylistic point… Was there a reason for using grep
> > for the various tests against system/distro strings? Why not a simple
> > case
> > statement, or even glob matching within [[ … ]]? I wondered whether
> > perhaps the need to run "type sw_vers" led you to convert the case/esac
> > to
> > an if/else, but even there, glob matching would have sufficed. It's not a
> > big deal, but there's something about spawning processes for things that
> > bash (or even sh) can handle natively that perturbs my aesthetic
> > sensibilities… ;-)
> > 
> > Thanks,
> > Brett S.
> > On Nov 11, 2015 4:33 PM, "Doug Birdwell" notifications@github.com
> > wrote:
> > 
> > > Brett,
> > > 
> > > I have edited the build-cac-enabled-git script to use a different
> > > method
> > > to identify the OS and modified the list of required packages so that
> > > everything builds on, at least, Debian, Ubuntu, CentOS, and Fedora.
> > > I have
> > > not changed the README.md file -- that should be (slightly) updated.
> > > 
> > > I have tested all the steps EXCEPT steps 3 & 10. I haven't tested 3 &
> > > 10
> > > because I'm still waiting for a CAC (!). I didn't test on Red Hat
> > > Enterprise Linux b/c I don't have a key, but it and CentOS should
> > > behave
> > > the same. I also didn't test on Cygwin (again, no system to test).
> > > If you
> > > still have access to a Cygwin system it would be good to test that & be
> > > sure I didn't mess anything up -- I didn't change (intentionally) the
> > > Cygwin-specific code. The behavior on Ubuntu should also be the same as
> > > before. Perhaps someone else can check the RHEL build (or I might be
> > > able
> > > to impose on a friend...).
> > > 
> > > The test for a Mac OS X system is for future use -- the problem there
> > > being the complexity of the installer and absence of a uniform set of
> > > pre-built packages. Any systems other than those listed are flagged as
> > > "unknown". Both Mac OS X and unknown cause graceful exits. Other
> > > Linux/UNIX
> > > systems can easily be added (with some experimentation to determine
> > > which
> > > packages to install and, possibly, the installer to use) -- the
> > > approach I
> > > am using to identify the OS is quite general. See
> > > http://linuxmafia.com/faq/Admin/release-files.html for more
> > > information.
> > > 
> > > Once I get a CAC & a CAC-enabled server I can point to, I will test
> > > steps
> > > 3 & 10, and execution. I'm using docker for all the tests. (Example,
> > > once
> > > you have the docker service installed and running: "docker run -ti -v
> > > /absolute-path-to-local-dir:/absolute-path-to-dir-in-container
> > > ubuntu:latest /bin/bash", followed by "apt-get update; apt-get install
> > > build-essential" in the container, followed by ./build-cac-enabled-git
> > > {options}. Substitute debian:latest, centos:latest, rhel:latest, or
> > > fedora:latest as needed, and the appropriate installer -- note that
> > > rhel
> > > requires a key before it will allow yum to install things.) I used
> > > docker
> > > and this approach in part because I wanted to be sure the build
> > > works when
> > > starting with a minimal development environment.
> > > 
> > > Note that the script does not check to ensure that the
> > > development/build
> > > tools have been installed -- the previous version doesn't do this,
> > > either.
> > > For Debian/Ubuntu this can be done with "apt-get install
> > > build-essential";
> > > for CentOS/Fedora/RHEL the command is "yum groupinstall 'Development
> > > Tools'" (or dnf instead of yum if dnf exists). "apt-get/yum/dnf update"
> > > should be executed before this.
> > > 
> > > One more thing to attempt in the future is to use apt-get/yum/dnf to
> > > install OpenSC rather than build by source. The OpenSC folks replied
> > > to my
> > > issue and said they have no control over the old versions associated
> > > with
> > > the various distributions.
> > > 
> > > ## Doug
> > > 
> > > You can view, comment on, or merge this pull request online at:
> > > 
> > > https://github.com/bpstahlman/cac-enabled-git-setup/pull/5
> > > Commit Summary
> > > - add multi-OS support and cleaned up package dependencies
> > > 
> > > File Changes
> > > - _M_ build-cac-enabled-git
> > > 
> > > <https://github.com/bpstahlman/cac-enabled-git-setup/pull/5/files#diff-0
> > > 
> > > (140)
> > > 
> > > Patch Links:
> > > - https://github.com/bpstahlman/cac-enabled-git-setup/pull/5.patch
> > > - https://github.com/bpstahlman/cac-enabled-git-setup/pull/5.diff
> > > 
> > > —
> > > Reply to this email directly or view it on GitHub
> > > https://github.com/bpstahlman/cac-enabled-git-setup/pull/5.
> > 
> > —
> > Reply to this email directly or view it on GitHub
> > <
> > https://github.com/bpstahlman/cac-enabled-git-setup/pull/5#issuecomment-157925720
> > .
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/bpstahlman/cac-enabled-git-setup/pull/5#issuecomment-157935735
> .
